data_steward/cdr_cleaner/cleaning_rules/generate_derived_tables.py

```diff
@@ -0,0 +1,379 @@
+,cteConditionEnds AS (
+  SELECT c.person_id,c.condition_concept_id,c.condition_start_date,MIN(e.end_date) AS era_end_date
+  FROM cteConditionTarget c
+  JOIN cteCondEndDates e ON c.person_id = e.person_id
+    AND c.condition_concept_id = e.condition_concept_id
+    AND e.end_date >= c.condition_start_date
+  GROUP BY c.person_id,c.condition_concept_id,c.condition_start_date)
+SELECT ROW_NUMBER() OVER (ORDER BY person_id) AS condition_era_id
+  ,person_id
+  ,condition_concept_id
+  ,TIMESTAMP(MIN(condition_start_date)) AS condition_era_start_date
+  ,TIMESTAMP(era_end_date) AS condition_era_end_date
+  ,COUNT(*) AS condition_occurrence_count
+FROM cteConditionEnds
+GROUP BY person_id,condition_concept_id,era_end_date
+""")
+
+
+class CreateDerivedTables(BaseCleaningRule):
+    """
+    Create derived tables.
+    """
+
+    def __init__(self,
+                 project_id,
+                 dataset_id,
+                 sandbox_dataset_id,
+                 ehr_cutoff_date=None,
+                 table_namer=None):
+        desc = 'Create derived tables post-deid in both CT/RT and both base/clean.'
+
+        try:
+            # set to provided date string if the date string is valid
+            self.ehr_cutoff_date = validate_date_string(ehr_cutoff_date)
+        except (TypeError, ValueError):
+            # otherwise, default to using today's date as the date string
+            self.cutoff_date = str(datetime.now().date())
+
+        super().__init__(issue_numbers=['DC3729'],
+                         description=desc,
+                         affected_datasets=[
+                             cdr_consts.REGISTERED_TIER_DEID_BASE,
+                             cdr_consts.REGISTERED_TIER_DEID_CLEAN,
+                             cdr_consts.CONTROLLED_TIER_DEID_BASE,
+                             cdr_consts.CONTROLLED_TIER_DEID_CLEAN
+                         ],
+                         project_id=project_id,
+                         dataset_id=dataset_id,
+                         sandbox_dataset_id=sandbox_dataset_id,
+                         depends_on=[],
+                         table_namer=table_namer)
+
+    def get_query_specs(self):
+        """
+        :return: a list of SQL strings to run
+        """
+        create_sandbox_table_list = []
+        for table in [OBSERVATION_PERIOD, DRUG_ERA, CONDITION_ERA]:
+            create_sandbox_table_list.append({
+                cdr_consts.QUERY:
+                    DELETION_QUERY.render(project_id=self.project_id,
+                                          dataset_id=self.dataset_id,
+                                          storage_table_name=table)
+            })
+        create_observation_period_table = POPULATE_OBS_PRD.render(
+            project_id=self.project_id,
+            dataset_id=self.dataset_id,
+            sandbox_dataset_id=self.sandbox_dataset_id,
+            ehr_cutoff_date=self.ehr_cutoff_date,
+            storage_table_name=OBSERVATION_PERIOD)
```

Comment:
Add cols, and fields can be retrieved from the schema json
```suggestion
            storage_table_name=OBSERVATION_PERIOD,
            cols=f"({','.join(fields_list)})")
```